File: src/main/java/net/ltgt/gradle/errorprone/ErrorPronePlugin.java
package net.ltgt.gradle.errorprone;

import static java.util.Collections.emptyList;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.gradle.api.Action;
import org.gradle.api.JavaVersion;
import org.gradle.api.Named;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.gradle.api.artifacts.Configuration;
import org.gradle.api.plugins.ExtensionAware;
import org.gradle.api.plugins.JavaBasePlugin;
import org.gradle.api.provider.Provider;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.Internal;
import org.gradle.api.tasks.Nested;
import org.gradle.api.tasks.Optional;
import org.gradle.api.tasks.SourceSet;
import org.gradle.api.tasks.SourceSetContainer;
import org.gradle.api.tasks.compile.CompileOptions;
import org.gradle.api.tasks.compile.JavaCompile;
import org.gradle.jvm.toolchain.JavaCompiler;
import org.gradle.process.CommandLineArgumentProvider;
import org.gradle.util.GradleVersion;
import org.jspecify.annotations.Nullable;

/**
 * A {@link Plugin} that configures {@link JavaCompile} tasks to use the <a
 * href="https://errorprone.info/">Error Prone compiler</a>.
 */
public abstract class ErrorPronePlugin implements Plugin<Project> {
  public static final String PLUGIN_ID = "net.ltgt.errorprone";
  public static final String CONFIGURATION_NAME = "errorprone";

  static final String EXTENSION_NAME = "errorprone";

  static final String TOO_OLD_TOOLCHAIN_ERROR_MESSAGE =
      "Must not enable ErrorProne when compiling with JDK < 11";

  static final String TEST_SOURCE_SET_NAME_REGEX = "^(t|.*T)est(\\p{javaUpperCase}.*)?$";

  static final List<String> JVM_ARGS_STRONG_ENCAPSULATION =
      Collections.unmodifiableList(
          Arrays.asList(
              "--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED",
              "--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED",
              "--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED",
              "--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED",
              "--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED",
              "--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED",
              "--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED",
              "--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED",
              "--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED",
              "--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED"));

  @Override
  public void apply(Project project) {
    if (GradleVersion.current().compareTo(GradleVersion.version("7.1")) < 0) {
      throw new UnsupportedOperationException("net.ltgt.errorprone requires at least Gradle 7.1");
    }

    Provider<Configuration> errorproneConfiguration = registerConfiguration(project);
    configureJavaCompileTasks(project);
    project
        .getPlugins()
        .withType(
            JavaBasePlugin.class,
            unused -> configureForJavaPlugin(project, errorproneConfiguration));
  }

  @SuppressWarnings("deprecation") // Configuration.setVisible for backwards-compatibility
  private Provider<Configuration> registerConfiguration(Project project) {
    return project
        .getConfigurations()
        .register(
            CONFIGURATION_NAME,
            configuration -> {
              configuration.setDescription(
                  "Error Prone dependencies, will be extended by all source sets' annotationProcessor configurations");
              configuration.setVisible(false);
              configuration.setCanBeConsumed(false);
              configuration.setCanBeResolved(false);

              Map<String, String> exclude = new HashMap<>(2);
              exclude.put("group", "com.google.errorprone");
              exclude.put("module", "javac");
              configuration.exclude(exclude);
            });
  }

  private void configureJavaCompileTasks(Project project) {
    project
        .getTasks()
        .withType(JavaCompile.class)
        .configureEach(this::configureJavaCompileTaskDefaults);
  }

  private void configureJavaCompileTaskDefaults(JavaCompile task) {
    ErrorProneOptions errorproneOptions =
        ((ExtensionAware) task.getOptions())
            .getExtensions()
            .create(EXTENSION_NAME, ErrorProneOptions.class);

    task.getOptions()
        .getCompilerArgumentProviders()
        .add(new ErrorProneCompilerArgumentProvider(errorproneOptions));

    ErrorProneJvmArgumentProvider jvmArgumentProvider =
        new ErrorProneJvmArgumentProvider(task, errorproneOptions);
    task.getOptions().getForkOptions().getJvmArgumentProviders().add(jvmArgumentProvider);
    task.doFirst(
        "Configure forking for errorprone",
        new ConfigureForkingTaskAction(errorproneOptions, jvmArgumentProvider, task.getOptions()));
  }

  private void configureForJavaPlugin(
      Project project, Provider<Configuration> errorproneConfiguration) {
    project
        .getExtensions()
        .getByType(SourceSetContainer.class)
        .configureEach(
            sourceSet -> configureForSourceSet(project, errorproneConfiguration, sourceSet));
  }

  private void configureForSourceSet(
      Project project, Provider<Configuration> errorproneConfiguration, SourceSet sourceSet) {
    project
        .getConfigurations()
        .named(
            sourceSet.getAnnotationProcessorConfigurationName(),
            configuration -> configuration.extendsFrom(errorproneConfiguration.get()));
    project
        .getTasks()
        .named(
            sourceSet.getCompileJavaTaskName(),
            JavaCompile.class,
            task -> configureTaskForSourceSet(sourceSet, task));
  }

  private void configureTaskForSourceSet(SourceSet sourceSet, JavaCompile task) {
    ((ExtensionAware) task.getOptions())
        .getExtensions()
        .configure(
            ErrorProneOptions.class,
            errorproneOptions ->
                configureSourceSetCompileJavaTask(sourceSet, task, errorproneOptions));
  }

  private void configureSourceSetCompileJavaTask(
      SourceSet sourceSet, JavaCompile task, ErrorProneOptions errorproneOptions) {
    errorproneOptions
        .getEnabled()
        .convention(
            task.getJavaCompiler()
                .map(
                    javaCompile ->
                        javaCompile.getMetadata().getLanguageVersion().canCompileOrRun(11))
                .orElse(true));
    errorproneOptions
        .getCompilingTestOnlyCode()
        .convention(sourceSet.getName().matches(TEST_SOURCE_SET_NAME_REGEX));
  }

  private static class ErrorProneCompilerArgumentProvider
      implements CommandLineArgumentProvider, Named {
    private final ErrorProneOptions errorproneOptions;

    ErrorProneCompilerArgumentProvider(ErrorProneOptions errorproneOptions) {
      this.errorproneOptions = errorproneOptions;
    }

    @Internal
    @Override
    public String getName() {
      return EXTENSION_NAME;
    }

    @SuppressWarnings("unused")
    @Nested
    @Optional
    @Nullable ErrorProneOptions getErrorproneOptions() {
      return errorproneOptions.getEnabled().getOrElse(false) ? errorproneOptions : null;
    }

    @Override
    public Iterable<String> asArguments() {
      if (!errorproneOptions.getEnabled().getOrElse(false)) {
        return emptyList();
      }
      return Arrays.asList(
          "-Xplugin:ErrorProne " + errorproneOptions,
          "-XDcompilePolicy=simple",
          "--should-stop=ifError=FLOW",
          // Error Prone 2.46.0 requires it for JDK 21 (and it helps NullAway too even with previous
          // Error Prone versions)
          // It's only useful for JDK 21, but safe to pass to any version.
          // See https://github.com/google/error-prone/issues/5426
          "-XDaddTypeAnnotationsToSymbol=true");
    }
  }

  private static class ErrorProneJvmArgumentProvider implements CommandLineArgumentProvider, Named {
    private final JavaCompile task;
    private final ErrorProneOptions errorproneOptions;

    ErrorProneJvmArgumentProvider(JavaCompile task, ErrorProneOptions errorproneOptions) {
      this.task = task;
      this.errorproneOptions = errorproneOptions;
    }

    @Internal
    @Override
    public String getName() {
      return EXTENSION_NAME;
    }

    @Input
    @Optional
    @Nullable JavaVersion getCompilerVersion() {
      JavaCompiler javaCompiler = task.getJavaCompiler().getOrNull();
      if (javaCompiler == null) {
        return isCommandLine(task.getOptions()) ? null : JavaVersion.current();
      }
      return JavaVersion.toVersion(javaCompiler.getMetadata().getLanguageVersion().asInt());
    }

    private boolean isCommandLine(CompileOptions options) {
      return options.isFork()
          && (options.getForkOptions().getJavaHome() != null
              || options.getForkOptions().getExecutable() != null);
    }

    @Override
    public Iterable<String> asArguments() {
      if (!errorproneOptions.getEnabled().getOrElse(false)) {
        return emptyList();
      }
      JavaVersion compilerVersion = getCompilerVersion();
      if (compilerVersion == null || !compilerVersion.isCompatibleWith(JavaVersion.VERSION_11)) {
        return emptyList();
      }
      return JVM_ARGS_STRONG_ENCAPSULATION;
    }
  }

  private static class ConfigureForkingTaskAction implements Action<Task> {
    private final ErrorProneOptions errorproneOptions;
    private final ErrorProneJvmArgumentProvider jvmArgumentProvider;
    private final CompileOptions options;

    ConfigureForkingTaskAction(
        ErrorProneOptions errorproneOptions,
        ErrorProneJvmArgumentProvider jvmArgumentProvider,
        CompileOptions options) {
      this.errorproneOptions = errorproneOptions;
      this.jvmArgumentProvider = jvmArgumentProvider;
      this.options = options;
    }

    @Override
    public void execute(Task unused) {
      if (!errorproneOptions.getEnabled().getOrElse(false)) {
        return;
      }
      JavaVersion compilerVersion = jvmArgumentProvider.getCompilerVersion();
      if (compilerVersion == null) {
        return;
      }
      if (!compilerVersion.isCompatibleWith(JavaVersion.VERSION_11)) {
        throw new UnsupportedOperationException(TOO_OLD_TOOLCHAIN_ERROR_MESSAGE);
      }
      if (!options.isFork()
          && compilerVersion.equals(JavaVersion.current())
          && StrongEncapsulationHelperJava.CURRENT_JVM_NEEDS_FORKING) {
        options.setFork(true);
      }
    }
  }

  private static class StrongEncapsulationHelperJava {
    static final boolean CURRENT_JVM_NEEDS_FORKING = currentJvmNeedsForking();

    private static boolean currentJvmNeedsForking() {
      if (!JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_16)) {
        return false;
      }
      try {
        // Only needed because we compile with --release 8
        // XXX: remove reflection when bumping the minimum required JDK, or using a Multi-Release
        // JAR
        Method getUnnamedModule = ClassLoader.class.getMethod("getUnnamedModule");
        Method getModule = Class.class.getMethod("getModule");
        Method getPackageName = Class.class.getMethod("getPackageName");
        Class<?> moduleClass = getModule.getReturnType();
        Method isExported = moduleClass.getMethod("isExported", String.class, moduleClass);
        Method isOpen = moduleClass.getMethod("isOpen", String.class, moduleClass);

        Object unnamedModule =
            getUnnamedModule.invoke(StrongEncapsulationHelperJava.class.getClassLoader());
        for (String className :
            new String[] {
              "com.sun.tools.javac.api.BasicJavacTask",
              "com.sun.tools.javac.api.JavacTrees",
              "com.sun.tools.javac.file.JavacFileManager",
              "com.sun.tools.javac.main.JavaCompiler",
              "com.sun.tools.javac.model.JavacElements",
              "com.sun.tools.javac.parser.JavacParser",
              "com.sun.tools.javac.processing.JavacProcessingEnvironment",
              "com.sun.tools.javac.tree.JCTree",
              "com.sun.tools.javac.util.JCDiagnostic",
            }) {
          Class<?> klass = Class.forName(className);
          if (Boolean.FALSE.equals(
              isExported.invoke(
                  getModule.invoke(klass), getPackageName.invoke(klass), unnamedModule))) {
            return true;
          }
        }
        for (String className :
            new String[] {
              "com.sun.tools.javac.code.Symbol", //
              "com.sun.tools.javac.comp.Enter",
            }) {
          Class<?> klass = Class.forName(className);
          if (Boolean.FALSE.equals(
              isOpen.invoke(
                  getModule.invoke(klass), getPackageName.invoke(klass), unnamedModule))) {
            return true;
          }
        }
      } catch (ClassNotFoundException ignored) {
        return true;
      } catch (InvocationTargetException | IllegalAccessException | NoSuchMethodException e) {
        throw new LinkageError("Shouldn't happen", e);
      }
      return false;
    }
  }
}

File: src/main/kotlin/net/ltgt/gradle/errorprone/ErrorPronePlugin.kt
@file:Suppress("ktlint:standard:no-wildcard-imports")

package net.ltgt.gradle.errorprone

import org.gradle.api.JavaVersion
import org.gradle.api.Named
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.file.FileCollection
import org.gradle.api.plugins.ExtensionAware
import org.gradle.api.plugins.JavaBasePlugin
import org.gradle.api.provider.ProviderFactory
import org.gradle.api.tasks.Classpath
import org.gradle.api.tasks.ClasspathNormalizer
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.Internal
import org.gradle.api.tasks.Nested
import org.gradle.api.tasks.Optional
import org.gradle.api.tasks.SourceSetContainer
import org.gradle.api.tasks.compile.CompileOptions
import org.gradle.api.tasks.compile.JavaCompile
import org.gradle.kotlin.dsl.*
import org.gradle.process.CommandLineArgumentProvider
import org.gradle.util.GradleVersion
import javax.inject.Inject

/**
 * A [Plugin] that configures [JavaCompile] tasks to use the [Error Prone compiler](https://errorprone.info/).
 */
class ErrorPronePlugin
    @Inject
    constructor(
        private val providers: ProviderFactory,
    ) : Plugin<Project> {
        companion object {
            const val PLUGIN_ID = "net.ltgt.errorprone"

            const val CONFIGURATION_NAME = "errorprone"

            const val JAVAC_CONFIGURATION_NAME = "errorproneJavac"

            internal const val TOO_OLD_TOOLCHAIN_ERROR_MESSAGE = "Must not enable ErrorProne when compiling with JDK < 8"

            private val HAS_JVM_ARGUMENT_PROVIDERS = GradleVersion.current() >= GradleVersion.version("7.1")

            internal val JVM_ARGS_STRONG_ENCAPSULATION =
                listOf(
                    "--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED",
                    "--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED",
                    "--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED",
                    "--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED",
                    "--add-exports=jdk.compiler/com.sun.tools.javac.parser=ALL-UNNAMED",
                    "--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED",
                    "--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED",
                    "--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED",
                    "--add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED",
                    "--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED",
                )

            private val CURRENT_JVM_NEEDS_FORKING by lazy {
                // Needs bootclasspath
                JavaVersion.current() == JavaVersion.VERSION_1_8 ||
                    (
                        // Needs --add-exports and --add-opens
                        JavaVersion.current() >= JavaVersion.VERSION_16 &&
                            StrongEncapsulationHelper().needsForking()
                    )
            }
        }

        override fun apply(project: Project) {
            if (GradleVersion.current() < GradleVersion.version("6.8")) {
                throw UnsupportedOperationException("$PLUGIN_ID requires at least Gradle 6.8")
            }

            val errorproneConfiguration =
                project.configurations.create(CONFIGURATION_NAME) {
                    description = "Error Prone dependencies, will be extended by all source sets' annotationProcessor configurations"
                    isVisible = false
                    isCanBeConsumed = false
                    isCanBeResolved = false

                    exclude(group = "com.google.errorprone", module = "javac")
                }
            val javacConfiguration: FileCollection =
                project.configurations.create(JAVAC_CONFIGURATION_NAME) {
                    description = "Error Prone Javac dependencies, will only be used when using JDK 8 (i.e. not JDK 9 or superior)"
                    isVisible = false
                    isCanBeConsumed = false
                    isCanBeResolved = true
                    defaultDependencies {
                        add(project.dependencies.create("com.google.errorprone:javac:9+181-r4173-1"))
                    }
                }

            project.tasks.withType<JavaCompile>().configureEach {
                val errorproneOptions =
                    (options as ExtensionAware).extensions.create(ErrorProneOptions.NAME, ErrorProneOptions::class.java)
                options
                    .compilerArgumentProviders
                    .add(ErrorProneCompilerArgumentProvider(errorproneOptions))

                val jvmArgumentProvider = ErrorProneJvmArgumentProvider(this, errorproneOptions, javacConfiguration)
                if (HAS_JVM_ARGUMENT_PROVIDERS) {
                    options.forkOptions.jvmArgumentProviders.add(jvmArgumentProvider)
                } else {
                    inputs
                        .property("errorprone.compilerVersion", providers.provider { jvmArgumentProvider.compilerVersion })
                        .optional(true)
                    inputs
                        .files(providers.provider { jvmArgumentProvider.bootstrapClasspath })
                        .withPropertyName("errorprone.bootstrapClasspath")
                        .withNormalizer(ClasspathNormalizer::class)
                        .optional()
                    doFirst("Configure JVM arguments for errorprone") {
                        options.forkOptions.jvmArgs!!.addAll(jvmArgumentProvider.asArguments())
                    }
                }
                doFirst("Configure forking for errorprone") {
                    if (!errorproneOptions.isEnabled.getOrElse(false)) return@doFirst
                    jvmArgumentProvider.compilerVersion?.let {
                        if (it < JavaVersion.VERSION_1_8) throw UnsupportedOperationException(TOO_OLD_TOOLCHAIN_ERROR_MESSAGE)
                        if ((it == JavaVersion.VERSION_1_8 || (it == JavaVersion.current() && CURRENT_JVM_NEEDS_FORKING)) &&
                            !options.isFork
                        ) {
                            options.isFork = true
                        }
                    }
                }
            }

            project.plugins.withType<JavaBasePlugin> {
                project.extensions.getByName<SourceSetContainer>("sourceSets").configureEach {
                    project.configurations[annotationProcessorConfigurationName].extendsFrom(errorproneConfiguration)
                    project.tasks.named<JavaCompile>(compileJavaTaskName) {
                        options.errorprone {
                            isEnabled.convention(javaCompiler.map { it.metadata.languageVersion.asInt() >= 8 }.orElse(true))
                            isCompilingTestOnlyCode.convention(this@configureEach.name.matches(TEST_SOURCE_SET_NAME_REGEX))
                        }
                    }
                }
            }
        }
    }

internal class ErrorProneJvmArgumentProvider(
    private val task: JavaCompile,
    private val errorproneOptions: ErrorProneOptions,
    private val javacConfiguration: FileCollection,
) : CommandLineArgumentProvider,
    Named {
    @Internal override fun getName(): String = "errorprone"

    @get:Input
    @get:Optional
    val compilerVersion by lazy {
        task.javaCompiler
            .map { JavaVersion.toVersion(it.metadata.languageVersion.asInt()) }
            .orNull ?: if (task.options.isCommandLine) null else JavaVersion.current()
    }

    @get:Classpath
    @get:Optional
    val bootstrapClasspath get() =
        javacConfiguration.takeIf {
            errorproneOptions.isEnabled.getOrElse(false) &&
                compilerVersion == JavaVersion.VERSION_1_8
        }

    override fun asArguments(): Iterable<String> =
        when {
            !errorproneOptions.isEnabled.getOrElse(false) -> emptyList()
            compilerVersion == null -> emptyList()
            compilerVersion == JavaVersion.VERSION_1_8 -> listOf("-Xbootclasspath/p:${javacConfiguration.asPath}")
            compilerVersion!! > JavaVersion.VERSION_1_8 -> ErrorPronePlugin.JVM_ARGS_STRONG_ENCAPSULATION
            else -> emptyList()
        }
}

internal class ErrorProneCompilerArgumentProvider(
    private val errorproneOptions: ErrorProneOptions,
) : CommandLineArgumentProvider,
    Named {
    @Internal override fun getName(): String = "errorprone"

    @Suppress("unused")
    @Nested
    @Optional
    fun getErrorproneOptions(): ErrorProneOptions? = errorproneOptions.takeIf { it.isEnabled.getOrElse(false) }

    override fun asArguments(): Iterable<String> =
        when {
            // should-stop.ifError is for JDK 9+, shouldStopPolicyIfError for JDK 8; it's safe to indiscriminately pass both
            errorproneOptions.isEnabled.getOrElse(
                false,
            ) ->
                listOf(
                    "-Xplugin:ErrorProne $errorproneOptions",
                    "-XDcompilePolicy=simple",
                    "-XDshould-stop.ifError=FLOW",
                    "-XDshouldStopPolicyIfError=FLOW",
                )
            else -> emptyList()
        }
}

internal val TEST_SOURCE_SET_NAME_REGEX =
    """^(t|.*T)est(\p{javaUpperCase}.*)?$""".toRegex()

private val CompileOptions.isCommandLine
    @Suppress("DEPRECATION")
    get() = isFork && (forkOptions.javaHome != null || forkOptions.executable != null)

private class StrongEncapsulationHelper {
    fun needsForking() =
        try {
            val unnamedModule: Any = this::class.java.classLoader.unnamedModule
            sequenceOf(
                "com.sun.tools.javac.api.BasicJavacTask",
                "com.sun.tools.javac.api.JavacTrees",
                "com.sun.tools.javac.file.JavacFileManager",
                "com.sun.tools.javac.main.JavaCompiler",
                "com.sun.tools.javac.model.JavacElements",
                "com.sun.tools.javac.parser.JavacParser",
                "com.sun.tools.javac.processing.JavacProcessingEnvironment",
                "com.sun.tools.javac.tree.JCTree",
                "com.sun.tools.javac.util.JCDiagnostic",
            ).any {
                val klass = Class.forName(it)
                return@any !klass.module.isExported(klass.packageName, unnamedModule)
            } &&
                sequenceOf(
                    "com.sun.tools.javac.code.Symbol",
                    "com.sun.tools.javac.comp.Enter",
                ).any {
                    val klass = Class.forName(it)
                    return@any !klass.module.isOpen(klass.packageName, unnamedModule)
                }
        } catch (e: ClassNotFoundException) {
            true
        }

    // Defined for backward/forward compatibility:
    // - compiles with jdk-release=8 just fine
    // - just remove those when upgrading to a newer minimum JDK without having to touch the code above
    private val getPackageName: java.lang.reflect.Method = Class::class.java.getMethod("getPackageName")
    private val getModule: java.lang.reflect.Method = Class::class.java.getMethod("getModule")
    private val isExported: java.lang.reflect.Method =
        getModule.returnType.getMethod(
            "isExported",
            String::class.java,
            getModule.returnType,
        )
    private val isOpen: java.lang.reflect.Method = getModule.returnType.getMethod("isOpen", String::class.java, getModule.returnType)
    private val getUnnamedModule: java.lang.reflect.Method = ClassLoader::class.java.getMethod("getUnnamedModule")

    private val Class<*>.packageName get(): String = getPackageName(this) as String
    private val Class<*>.module get(): Any = getModule(this)

    private fun Any.isExported(
        pn: String,
        other: Any,
    ): Boolean = isExported(this, pn, other) as Boolean

    private fun Any.isOpen(
        pn: String,
        other: Any,
    ): Boolean = isOpen(this, pn, other) as Boolean

    private val ClassLoader.unnamedModule: Any get() = getUnnamedModule(this)
}
